pkg/grafana/client/client.go

```diff
@@ -1,53 +1,105 @@
 package client
 
 import (
+	"context"
+	"crypto/tls"
 	"fmt"
 	"net/url"
 
 	grafana "github.com/grafana/grafana-openapi-client-go/client"
-
-	"github.com/giantswarm/observability-operator/pkg/config"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
 	clientConfigNumRetries = 3
+
+	// Secret names and keys for Grafana configuration
+	grafanaNamespace = "monitoring"
+
+	grafanaAdminSecretName        = "grafana"
+	grafanaAdminSecretUserKey     = "admin-user"
+	grafanaAdminSecretPasswordKey = "admin-password"
+
+	grafanaTLSSecretName    = "grafana-tls"
+	grafanaTLSSecretCertKey = "tls.crt"
+	grafanaTLSSecretKeyKey  = "tls.key"
 )
 
-func GenerateGrafanaClient(grafanaURL *url.URL, conf config.Config) (*grafana.GrafanaHTTPAPI, error) {
-	var err error
+// GenerateGrafanaClient creates a new Grafana client by fetching credentials
+// and TLS configuration from Kubernetes secrets.
+func GenerateGrafanaClient(
+	ctx context.Context,
+	k8sClient client.Client,
+	grafanaURL *url.URL,
+) (*grafana.GrafanaHTTPAPI, error) {
+	// Get Grafana admin credentials from secret
+	adminSecret := &corev1.Secret{}
+	adminSecretKey := client.ObjectKey{Namespace: grafanaNamespace, Name: grafanaAdminSecretName}
+	if err := k8sClient.Get(ctx, adminSecretKey, adminSecret); err != nil {
+		return nil, fmt.Errorf("failed to get Grafana admin secret %q in namespace %q: %w", grafanaAdminSecretName, grafanaNamespace, err)
+	}
 
-	// Generate Grafana client
-	// Get grafana admin-password and admin-user
-	adminUserCredentials := AdminCredentials{
-		Username: conf.Environment.GrafanaAdminUsername,
-		Password: conf.Environment.GrafanaAdminPassword,
+	adminUsernameBytes, ok := adminSecret.Data[grafanaAdminSecretUserKey]
+	if !ok {
+		return nil, fmt.Errorf("key %q not found in Grafana admin secret %q", grafanaAdminSecretUserKey, grafanaAdminSecretName)
+	}
+	if len(adminUsernameBytes) == 0 {
+		return nil, fmt.Errorf("GrafanaAdminUsername is empty in secret %q", grafanaAdminSecretName)
 	}
-	if adminUserCredentials.Username == "" {
-		return nil, fmt.Errorf("GrafanaAdminUsername not set: %q", conf.Environment.GrafanaAdminUsername)
 
+	adminPasswordBytes, ok := adminSecret.Data[grafanaAdminSecretPasswordKey]
+	if !ok {
+		return nil, fmt.Errorf("key %q not found in Grafana admin secret %q", grafanaAdminSecretPasswordKey, grafanaAdminSecretName)
+	}
+	if len(adminPasswordBytes) == 0 {
+		return nil, fmt.Errorf("GrafanaAdminPassword is empty in secret %q", grafanaAdminSecretName)
 	}
-	if adminUserCredentials.Password == "" {
-		return nil, fmt.Errorf("GrafanaAdminPassword not set: %q", conf.Environment.GrafanaAdminPassword)
 
+	adminUserCredentials := AdminCredentials{
+		Username: string(adminUsernameBytes),
+		Password: string(adminPasswordBytes),
 	}
```

Comment:
Can we rather create the needed struct right away instead of having an intermediate adminUserCredentials ?

```suggestion
	adminUserCredentials := url.UserPassword(string(adminUsernameBytes), string(adminPasswordBytes))
```